usr/opt/OSCX200/src/man/XI/XListDev.man
.\"
.\" *****************************************************************
.\" *                                                               *
.\" *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
.\" *                                                               *
.\" *   All Rights Reserved.  Unpublished rights  reserved  under   *
.\" *   the copyright laws of the United States.                    *
.\" *                                                               *
.\" *   The software contained on this media  is  proprietary  to   *
.\" *   and  embodies  the  confidential  technology  of  Digital   *
.\" *   Equipment Corporation.  Possession, use,  duplication  or   *
.\" *   dissemination of the software and media is authorized only  *
.\" *   pursuant to a valid written license from Digital Equipment  *
.\" *   Corporation.                                                *
.\" *                                                               *
.\" *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
.\" *   by the U.S. Government is subject to restrictions  as  set  *
.\" *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
.\" *   or  in  FAR 52.227-19, as applicable.                       *
.\" *                                                               *
.\" *****************************************************************
.\"
.\"
.\" HISTORY
.\"
.ds xL Programming with Xlib
.TH XListInputDevices 3X11 "Release 5" "X Version 11" "X FUNCTIONS"
.SH NAME
XListInputDevices, XFreeDeviceList \- list available input devices
.SH SYNTAX
XDeviceInfo *XListInputDevices\^(\^\fIdisplay\fP, \fIndevices_return\fP\^)
.br
      Display *\fIdisplay\fP\^;
.br
      int *\fIndevices_return\fP\^;
.sp
XFreeDeviceList\^(\^\fIlist\fP\^)
.br
      XDeviceInfo *\fIlist\fP\^;
.SH ARGUMENTS
.TP 12
.I display
Specifies the connection to the X server.
.TP 12
.I ndevices_return
Specifies a pointer to a variable where the number of available devices can
be returned.
.TP 12
.I list
Specifies the list of devices to free.  The \fIXFreeDeviceList\fP function
frees the list of available extension input devices.
.SH DESCRIPTION
The \fIXListInputDevices\fP
request lists the available extension input devices.  This list includes the
X pointer and X keyboard, any other input devices that are currently accessible
through the X server, and any input devices that are not currently accessible
through the X server but could be accessed if requested.
.P
Some server implementations may make all input devices available at the time
the server is initialized.  Others may wait until requested by a client to
access an input device.  In the latter case, it is possible that an input 
device will be listed as available at one time but not at another.
.P
For each input device available to the server, the XListInputDevices
request returns an XDeviceInfo structure.  That structure contains a
pointer to a list of structures, each of which contains information about 
one class of input supported by the device.

The XDeviceInfo structure is defined as follows:

.DS
.nf
typedef struct _XDeviceInfo
{
	XID		id;        
	Atom		type;
	char		*name;
	int		num_classes;
	int		use;
	XAnyClassPtr	inputclassinfo;
} XDeviceInfo;
.fi
.DE
.P
The id is a number in the range 0-128 that uniquely identifies 
the device.  It is assigned to the device when it is initialized by the server.
.P
The type field is of type Atom and indicates the nature
of the device.
.P
The name field contains a pointer to a null-terminated
string that corresponds to one of the defined device
types.  The name will correspond to one of the following
strings (defined in the header file \fIXI.h\fP:
.P
.DS
XI_MOUSE
XI_TABLET
XI_KEYBOARD
XI_TOUCHSCREEN
XI_TOUCHPAD
XI_BUTTONBOX
XI_BARCODE
XI_TRACKBALL
XI_QUADRATURE
XI_ID_MODULE
XI_ONE_KNOB
XI_NINE_KNOB
XI_KNOB_BOX
XI_SPACEBALL
XI_DATAGLOVE
XI_EYETRACKER
XI_CURSORKEYS
XI_FOOTMOUSE
.DE
.P
These names may be directly compared with the name field of the
XDeviceInfo structure, or used in an XInternAtom request to return
an atom that can be compared with the type field of the XDeviceInfo
structure.
.P
The num_classes field is a number in the
range 0-255 that specifies the number of input classes
supported by the device for which information is
returned by ListInputDevices.  Some input classes, such
as class Focus and class Proximity do not have any
information to be returned by ListInputDevices.
.P
The use field specifies how the device is currently
being used.  If the value is \fIIsXKeyboard\fP, the device is
currently being used as the X keyboard.  If the value
is \fIIsXPointer\rP, the device is currently being used as
the X pointer.  If the value is \fIIsXExtensionDevice\fP, the
device is available for use as an extension device.
.P
The inputclassinfo field contains a pointer to the first input-class
specific data.  The first two fields are common to all
classes.
.P
The class field is a number in the range 0-255.
It uniquely identifies the class of input for which
information is returned.  Currently defined classes
are KeyClass, ButtonClass, and ValuatorClass.
.P
The length field is a number in the range 0- 255.  
It specifies the number of bytes of data that are
contained in this input class.  The length includes the
class and length fields.
.P
The XKeyInfo structure describes the characteristics of the keys on the
device.  It is defined as follows:
.P
.DS
.nf
typedef struct _XKeyInfo {
	XID			class;
	int			length;
	unsigned short		min_keycode;
	unsigned short		max_keycode;
	unsigned short		num_keys;
} XKeyInfo;
.fi
.DE
.P
min_keycode is of type KEYCODE.  It specifies the
minimum keycode that the device will report.  The
minimum keycode will not be smaller than 8.
.P
max_keycode is of type KEYCODE.  It specifies the
maximum keycode that the device will report.  The
maximum keycode will not be larger than 255.
.P
num_keys specifies the number of keys that the device has.
.P
The XButtonInfo structure defines the characteristics of the buttons
on the device.  It is defined as follows:
.P
.DS
.nf
typedef struct _XButtonInfo {
	XID		class;
	int		length;
	short		num_buttons;
} XButtonInfo;
.fi
.DE
.P
num_buttons specifies the number of buttons that the device has.
.P
The XValuatorInfo structure defines the characteristics of the valuators
on the device.  It is defined as follows:
.P
.DE
.nf
typedef struct	_XValuatorInfo {
	XID			class;
	int			length;
	unsigned char		num_axes;
	unsigned char		mode;
	unsigned long		motion_buffer;
	XAxisInfoPtr		axes;
} XValuatorInfo;
.fi
.DS
num_axes contains the number of axes the device supports.
.P
mode is a constant that has one of the following
values: Absolute or Relative.  Some devices allow the
mode to be changed dynamically via the SetDeviceMode
request.
.P
motion_buffer_size is a cardinal number that specifies
the number of elements that can be contained in the
motion history buffer for the device.
.P
The axes field contains a pointer to an XAxisInfo structure.
.P
The XAxisInfo structure is defined as follows:
.P
.DS
.nf
typedef struct _XAxisInfo {
	int 	resolution;
	int 	min_value;
	int 	max_value;
} XAxisInfo;
.fi
.DE
.P
The resolution contains a number in counts/meter.
.P
The min_val field contains a number that specifies
the minimum value the device reports for this axis.
For devices whose mode is Relative, the min_val field
will contain 0.
.P
The max_val field contains a number that specifies
the maximum value the device reports for this axis.
For devices whose mode is Relative, the max_val field
will contain 0.
.P
To free the \fIXDeviceInfo\fP array created by \fIXListInputDevices\fP,
use \fIXFreeDeviceList\fP.
.SH DIAGNOSTICS
none.
.SH "SEE ALSO"
.br
\fI\*(xL\fP


usr/opt/OSCX200/src/extensions/server/xinput/xlistdev.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
/* $Header: /usr/sde/x11/rcs/x11/src/./extensions/server/xinput/xlistdev.c,v 1.2 91/12/15 12:42:16 devrcs Exp $ */

/************************************************************
Copyright (c) 1989 by Hewlett-Packard Company, Palo Alto, California, and the 
Massachusetts Institute of Technology, Cambridge, Massachusetts.

			All Rights Reserved

Permission to use, copy, modify, and distribute this software and its
documentation for any purpose and without fee is hereby granted,
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in
supporting documentation, and that the names of Hewlett-Packard or MIT not be
used in advertising or publicity pertaining to distribution of the
software without specific, written prior permission.

HEWLETT-PACKARD DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL
HEWLETT-PACKARD BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR
ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
SOFTWARE.

********************************************************/

/***********************************************************************
 *
 * Extension function to list the available input devices.
 *
 */

#define	 NEED_EVENTS
#define	 NEED_REPLIES
#include "X.h"				/* for inputstr.h    */
#include "Xproto.h"			/* Request macro     */
#include "inputstr.h"			/* DeviceIntPtr	     */
#include "XI.h"
#include "XIproto.h"

extern InputInfo inputInfo;
extern	int 	IReqCode;
extern	int	BadDevice;
extern	void	(*ReplySwapVector[256]) ();
DeviceIntPtr	LookupDeviceIntRec();

void		CopySwapKeyClass ();
void		CopySwapButtonClass ();
void		CopySwapValuatorClass ();
void		SizeDeviceInfo ();
void		ListDeviceInfo ();
void		AddOtherInputDevices ();

/***********************************************************************
 *
 * This procedure lists the input devices available to the server.
 *
 */

int
SProcXListInputDevices(client)
    register ClientPtr client;
    {
    register char n;

    REQUEST(xListInputDevicesReq);
    swaps(&stuff->length, n);
    return(ProcXListInputDevices(client));
    }

/***********************************************************************
 *
 * This procedure lists the input devices available to the server.
 *
 */

ProcXListInputDevices (client)
    register ClientPtr client;
    {
    xListInputDevicesReply	rep;
    int			numdevs;
    int 		namesize = 1;	/* need 1 extra byte for strcpy */
    int 		size = 0;
    int 		total_length;
    char		*devbuf;
    char		*classbuf;
    char		*namebuf;
    char		*savbuf;
    xDeviceInfo 	*dev;
    DeviceIntPtr 	d;
    void CopyDeviceName ();
    void CopySwapDevice ();

    REQUEST(xListInputDevicesReq);
    REQUEST_SIZE_MATCH(xListInputDevicesReq);

    rep.repType = X_Reply;
    rep.RepType = X_ListInputDevices;
    rep.length = 0;
    rep.sequenceNumber = client->sequence;

    AddOtherInputDevices ();
    numdevs = inputInfo.numDevices;

    for (d=inputInfo.devices; d; d=d->next)
	SizeDeviceInfo (d, &namesize, &size);
    for (d=inputInfo.off_devices; d; d=d->next)
	SizeDeviceInfo (d, &namesize, &size);

    total_length = numdevs * sizeof (xDeviceInfo) + size + namesize;
    devbuf = (char *) Xalloc (total_length);
    classbuf = devbuf + (numdevs * sizeof (xDeviceInfo));
    namebuf = classbuf + size;
    savbuf = devbuf;

    dev = (xDeviceInfoPtr) devbuf;
    for (d=inputInfo.devices; d; d=d->next,dev++)
        ListDeviceInfo (client, d, dev, &devbuf, &classbuf, &namebuf);
    for (d=inputInfo.off_devices; d; d=d->next,dev++)
        ListDeviceInfo (client, d, dev, &devbuf, &classbuf, &namebuf);

    rep.ndevices = numdevs;
    rep.length = (total_length + 3) >> 2;
    WriteReplyToClient (client, sizeof (xListInputDevicesReply), &rep);
    WriteToClient(client, total_length, savbuf);
    Xfree (savbuf);
    return Success;
    }

/***********************************************************************
 *
 * This procedure calculates the size of the information to be returned
 * for an input device.
 *
 */

void
SizeDeviceInfo (d, namesize, size)
    DeviceIntPtr d;
    int *namesize;
    int *size;
    {
    *namesize += strlen (d->name) + 1;
    if (d->key != NULL)
	*size += sizeof (xKeyInfo);
    if (d->button != NULL)
	*size += sizeof (xButtonInfo);
    if (d->valuator != NULL)
	*size += (sizeof(xValuatorInfo) + 
		d->valuator->numAxes * sizeof(xAxisInfo));
    }

/***********************************************************************
 *
 * This procedure lists information to be returned for an input device.
 *
 */

void
ListDeviceInfo (client, d, dev, devbuf, classbuf, namebuf)
    ClientPtr client;
    DeviceIntPtr d;
    xDeviceInfoPtr dev;
    char **devbuf;
    char **classbuf;
    char **namebuf;
    {
    CopyDeviceName (namebuf, d->name);
    CopySwapDevice (client, d, 0, devbuf);
    if (d->key != NULL)
	{
	CopySwapKeyClass(client, d->key, classbuf);
	dev->num_classes++;
	}
    if (d->button != NULL)
	{
	CopySwapButtonClass(client, d->button, classbuf);
	dev->num_classes++;
	}
    if (d->valuator != NULL)
	{
	CopySwapValuatorClass(client, d->valuator, classbuf);
	dev->num_classes++;
	}
    }

/***********************************************************************
 *
 * This procedure copies data to the DeviceInfo struct, swapping if necessary.
 *
 * We need the extra byte in the allocated buffer, because the trailing null
 * hammers one extra byte, which is overwritten by the next name except for
 * the last name copied.
 *
 */

void
CopyDeviceName (namebuf, name)
    char **namebuf;
    char *name;
    {
    char *nameptr = (char *) *namebuf;

    *nameptr++ = strlen (name);
    strcpy (nameptr, name);
    *namebuf += (strlen (name)+1);
    }

/***********************************************************************
 *
 * This procedure copies data to the DeviceInfo struct, swapping if necessary.
 *
 */

void
CopySwapDevice (client, d, num_classes, buf)
    register ClientPtr 	client;
    DeviceIntPtr	d;
    int			num_classes;
    char 		**buf;
    {
    register char 	n;
    xDeviceInfoPtr dev;

    dev = (xDeviceInfoPtr) *buf;

    dev->id = d->id;
    dev->type = d->type;
    dev->num_classes = num_classes;
    if (d == inputInfo.keyboard)
	dev->use = IsXKeyboard;
    else if (d == inputInfo.pointer)
	dev->use = IsXPointer;
    else
	dev->use = IsXExtensionDevice;
    if (client->swapped)
	{
	swapl(&dev->type, n);	/* macro - braces are required */
	}
    *buf += sizeof (xDeviceInfo);
    }

/***********************************************************************
 *
 * This procedure copies KeyClass information, swapping if necessary.
 *
 */

void
CopySwapKeyClass (client, k, buf)
    register ClientPtr 	client;
    KeyClassPtr 	k;
    char 		**buf;
    {
    register char 	n;
    xKeyInfoPtr 	k2;

    k2 = (xKeyInfoPtr) *buf;
    k2->class = KeyClass;
    k2->length = sizeof (xKeyInfo);
    k2->min_keycode = k->curKeySyms.minKeyCode;
    k2->max_keycode = k->curKeySyms.maxKeyCode;
    k2->num_keys = k2->max_keycode - k2->min_keycode + 1;
    if (client->swapped)
	{
	swaps(&k2->num_keys,n);
	}
    *buf += sizeof (xKeyInfo);
    }

/***********************************************************************
 *
 * This procedure copies ButtonClass information, swapping if necessary.
 *
 */

void
CopySwapButtonClass (client, b, buf)
    register ClientPtr 	client;
    ButtonClassPtr 	b;
    char 		**buf;
    {
    register char 	n;
    xButtonInfoPtr 	b2;

    b2 = (xButtonInfoPtr) *buf;
    b2->class = ButtonClass;
    b2->length = sizeof (xButtonInfo);
    b2->num_buttons = b->numButtons;
    if (client->swapped)
	{
	swaps(&b2->num_buttons,n);	/* macro - braces are required */
	}
    *buf += sizeof (xButtonInfo);
    }

/***********************************************************************
 *
 * This procedure copies ValuatorClass information, swapping if necessary.
 *
 */

void
CopySwapValuatorClass (client, v, buf)
    register ClientPtr 	client;
    ValuatorClassPtr 	v;
    char 		**buf;
    {
    int			j;
    register char 	n;
    xValuatorInfoPtr 	v2;
    AxisInfo 		*a;
    xAxisInfoPtr 	a2;

    v2 = (xValuatorInfoPtr) *buf;
    v2->class = ValuatorClass;
    v2->length = sizeof (xValuatorInfo) + v->numAxes *
	sizeof (xAxisInfo);
    v2->num_axes  = v->numAxes;
    v2->mode  = v->mode & DeviceMode;
    v2->motion_buffer_size  = v->numMotionEvents;
    if (client->swapped)
	{
	swapl(&v2->motion_buffer_size,n);
	}
    *buf += sizeof (xValuatorInfo);
    a = v->axes;
    a2 = (xAxisInfoPtr) *buf;
    for (j=0; j<v->numAxes; j++)
	{
	a2->min_value = a->min_value;
	a2->max_value = a->max_value;
	a2->resolution = a->resolution;
    	if (client->swapped)
	    {
	    swapl(&a2->min_value,n);
	    swapl(&a2->max_value,n);
	    swapl(&a2->resolution,n);
	    }
	a2++;
	a++;
	*buf += sizeof (xAxisInfo);
	}
    }

/***********************************************************************
 *
 * This procedure writes the reply for the XListInputDevices function,
 * if the client and server have a different byte ordering.
 *
 */

SRepXListInputDevices (client, size, rep)
    ClientPtr	client;
    int		size;
    xListInputDevicesReply	*rep;
    {
    register char n;

    swaps(&rep->sequenceNumber, n);
    swapl(&rep->length, n);
    WriteToClient(client, size, rep);
    }
